### PR TITLE
Fix HTML capture race conditions

### LINKED ISSUE #547: Inconsistent HTML printing
Quoting #531:

> I'm developing a project where I send orders from a APP directly to the kitchen printer in a restaurant, after the payment is confirmed. Everything is working fine but sometimes the "ticket" printed is huge and I can't understand the problem.
> 
> Take a look at this examples:
> On the right you have a regular ticked printed as expected.
> On the left a similar ticket is printed with a huge font.
>
> IMG_6831
> 
> I can reprint the same order multiple times and 1 in 20 tickets is printed in this weird format.
Anyone had a similar issue?
>
> The computer connected with the printer is brand new, running windows 10 and the printer is a TM-T88VI-111

Quoting @klabarge:

> @Read1337 I was able to reproduce your exact issue. I'm looking more into the logs now, and will let you know when we have more information.
>
> My environment:
> 
> **OS:** Windows 10
> **Printer:** Epson TM-T88VI
> **Java:** AdoptOpenJDK 11.0.4 64-bit
> **QZ Tray:** 2.0.11
> 
> I was also able to reproduce this to a virtual printer.
> 
> 1. Setup Signing
> 1. Set Printer Page Size
>    * Right click printer > Printer properties > Prererences > Advanced
>    * Set page size to "PostScript Custom Page Size", make size 80mm x 297mm
> 
> image
> 
> image

### PR DESCRIPTION
**Pull request by tresf**
Creating a PR because initial tests are passing.

Closes #547 #32

### CODE REVIEW

File: src/qz/printer/action/WebApp.java [hunk 1]
```diff
@@ -131,28 +141,30 @@ public WebApp() {
...
-            log.trace("Waiting for JavaFX..");
-            try { Thread.sleep(SLEEP); } catch(Exception ignore) {}
-        }
+        if (startupLatch.getCount() > 0) {
+            try {
+                log.trace("Waiting for JavaFX..");
 
-        if (!started.get()) {
-            throw new IOException("JavaFX did not start");
+                if (!startupLatch.await(1000, TimeUnit.MILLISECONDS)) {
```
Comment (tresf) [inline src/qz/printer/action/WebApp.java]: I think this should be `60, TimeUnit.SECONDS` to match our existing code of `TIMEOUT = 60; //total paused seconds before failing`

File: src/qz/printer/action/WebApp.java [hunk 2]
```diff
@@ -76,40 +72,43 @@ public void changed(ObservableValue<? extends Worker.State> ov, Worker.State old
...
-                        webView.setMinHeight(pageHeight * pageZoom);
-                        webView.setPrefHeight(pageHeight * pageZoom);
-                        webView.autosize();
-
-                        snap.playFromStart();
+                    public void run() {
+                        log.debug("Attempting image capture");
+
+                        SnapshotParameters sparams = new SnapshotParameters();
+                        sparams.setTransform(new Scale(pageZoom, pageZoom));
```
Comment (tresf) [inline src/qz/printer/action/WebApp.java]: On Java 8 and Java 11 using 4x zoom, it appears to just make the picture bigger without improving quality... :(

image

**Comment by tresf** [comment]
@bberenz Tests are looking great so far.

Some requests:
* Avoid the color white as a background color
* Test the bottom right pixel to make sure the render is complete

Here's Java 11 + Gluon on macOS.  I can't find an issue at a glance with any of the renders.

ezgif-2-7ffa84ff649b

**Comment by tresf** [comment]
I just realized that the zoom is still using the `SnapshotParameters` logic, which is not correct. https://github.com/qzind/tray/pull/580#discussion_r377469631 so my test results are not usable.

File: src/qz/printer/action/WebApp.java (hunk 1 above)
Comment (tresf) [reply src/qz/printer/action/WebApp.java]: @bberenz bump as I don't think this snippet was ever resolved.

Comment (bberenz) [reply src/qz/printer/action/WebApp.java]: I thought I addressed it at some point, but it must have gotten reverted in one of the rollbacks. Taken care of now.

**Comment by klabarge** [comment]
I sent over 300 print jobs and all print jobs completed. I didn't hit the race condition using this branch

However, I did notice that the HTML rendering is much more noticeable 

Screen Recording 2020-03-24 at 8 48 13 PM

**Comment by tresf** [comment]
> However, I did notice that the HTML rendering is much more noticeable

I tried to fix it by making the stage smaller, but it results in more partial snapshots.  I guess that won't improve until #576 is completed.  🤞

Merging.

**Pull request merged**